File: src/packages/skill-hermes/agentxp_skill_hermes/canonical.py
"""Canonical byte sequence + SHA-256 event id.

Matches ADR-003 D1 exactly: recursively sorted-keys, whitespace-free
JSON over the envelope with `id` and `sig` removed. Output MUST be
byte-identical to @serendip/protocol's `canonicalize` so that any
signed event produced here verifies under the TypeScript verifier.
"""

from __future__ import annotations

import hashlib
import json
from typing import Any

_EXCLUDED_KEYS = frozenset({"id", "sig"})


def _sorted_json(value: Any) -> str:
    # Primitives route through json.dumps with ensure_ascii=False; JS
    # JSON.stringify leaves non-ASCII characters unescaped, so Python
    # must follow suit to keep the byte sequence identical.
    if value is None or isinstance(value, bool) or isinstance(value, (int, float, str)):
        return json.dumps(value, ensure_ascii=False)
    if isinstance(value, list):
        return "[" + ",".join(_sorted_json(v) for v in value) + "]"
    if isinstance(value, dict):
        keys = sorted(value.keys())
        pairs = [json.dumps(k, ensure_ascii=False) + ":" + _sorted_json(value[k]) for k in keys]
        return "{" + ",".join(pairs) + "}"
    raise TypeError(f"canonical: unsupported type {type(value).__name__}")


def canonicalize(event: dict[str, Any]) -> str:
    """Return the canonical UTF-8 string for an event envelope.

    `id` and `sig` are stripped from the top level only; nested
    objects are passed through untouched (keys sorted at every depth).
    """
    filtered = {k: v for k, v in event.items() if k not in _EXCLUDED_KEYS}
    return _sorted_json(filtered)


def sha256_hex(text: str) -> str:
    """SHA-256 of the UTF-8 bytes of `text`, lowercase hex."""
    return hashlib.sha256(text.encode("utf-8")).hexdigest()

File: packages/protocol/src/canonical.ts
// Canonical byte sequence and SHA-256 event id, per ADR-003 D1.
// Ported from legacy/src-v1/packages/protocol/src/events.ts:53-73.
import { sha256 } from '@noble/hashes/sha256'
import type { SerendipEvent } from './types.js'
import { bytesToHex } from './utils.js'

// Keys excluded from the canonical byte sequence. Their presence or
// absence on the input object does not affect the output.
const EXCLUDED_KEYS: ReadonlySet<string> = new Set(['id', 'sig'])

/**
 * Produce the canonical whitespace-free JSON string used as the
 * pre-image of `event.id`. Keys are sorted lexicographically at every
 * depth; `id` and `sig` are stripped from the top level only.
 *
 * `event` MAY be a complete SerendipEvent or any partial shape during
 * signing; extra fields that aren't part of the envelope are included
 * unchanged so that callers cannot smuggle un-signed data through.
 */
export function canonicalize(event: Partial<SerendipEvent>): string {
  const filtered: Record<string, unknown> = {}
  for (const [key, value] of Object.entries(event)) {
    if (EXCLUDED_KEYS.has(key)) continue
    filtered[key] = value
  }
  return sortedJson(filtered)
}

/**
 * Compute SHA-256 of the UTF-8 bytes of `input` and return it as
 * lowercase hex (64 characters).
 */
export function sha256hex(input: string): string {
  const bytes = new TextEncoder().encode(input)
  return bytesToHex(sha256(bytes))
}

function sortedJson(value: unknown): string {
  if (value === null || typeof value !== 'object') {
    return JSON.stringify(value)
  }
  if (Array.isArray(value)) {
    return '[' + value.map(sortedJson).join(',') + ']'
  }
  const obj = value as Record<string, unknown>
  const keys = Object.keys(obj).sort()
  const pairs: string[] = []
  for (const key of keys) {
    pairs.push(JSON.stringify(key) + ':' + sortedJson(obj[key]))
  }
  return '{' + pairs.join(',') + '}'
}
